_episodes/l1-03-intermediate_concepts.md

```diff
@@ -1,21 +1,441 @@
+> > | *   2c2d0e2 Merge branch 'experiment'
+> > | |\
+> > | |/
+> > |/|
+> > * | d9043d2 Try with some coriander
+> > | * 6a2a76f Corrected typo in ingredients.md
+> > |/
+> > * 57d4505 Revert "Added instruction to enjoy"
+> > * 5cb4883 Added 1/2 onion
+> > * 43536f3 Added instruction to enjoy
+> > * 745fb8b Adding ingredients and instructions
+> > ~~~
+> > {: .output}
+> > Note that while the `experiment` branch still mentions the adjustment of salt, that is
+> > no longer part of the `main` commit history. Your working directory has become identical
+> > to before starting the salty that adventure.
+> >
+> {: .solution}
+>
+{: .challenge}
+> ## Changing History Can Have Unexpected Consequences
+>
+> Using `git reset` to remove a commit is a bad idea if
+> you have **shared** it yet with other people. If you make a commit and share it on
+> GitHub or with a colleague by other means then removing that commit from your Git
+> history will cause inconsistencies that may be difficult to resolve later. We
+> only recommend this approach for commits that are only in your local working
+> copy of a repository.
+{: .callout}
+
+
+### Removing branches once you are done with them is good practice
+Over time, you will accumulate lots of branches to implement different features in you
+code. It is good practice to remove them once they have fulfil their purpose. You can do
+that using the `-D` flag with the `git branch` command:
+
+~~~
+$ git branch -D BRANCH_NAME
+~~~
+{: .commands}
+
+> ## Getting rid of the experiment
+> As we are done with the `experiment` branch, let's delete it to have a cleaner history.
+> >
+> > ## Solution
+> > ~~~
+> > $ git branch -D experiment
+> > $ git graph
+> > ~~~
+> > {: .commands}
+> >
+> > Now, the commit history should look as:
+> > ~~~
+> > *   fe0d257 (HEAD -> main) Merge branch 'experiment'
+> > |\
+> > | * 99b2352 Reduced the amount of coriander
+> > * | 2c2d0e2 Merge branch 'experiment'
+> > |\|
+> > | * d9043d2 Try with some coriander
+> > * | 6a2a76f Corrected typo in ingredients.md
+> > |/
+> > * 57d4505 Revert "Added instruction to enjoy"
+> > * 5cb4883 Added 1/2 onion
+> > * 43536f3 Added instruction to enjoy
+> > * 745fb8b Adding ingredients and instructions
+> > ~~~
+> > {: .output}
+> >
+> > Now there is truly no trace of your attempts to change the content of salt!
+> >
+> {: .solution}
+>
+{: .challenge}
+### Reversing a commit
+
+As pointed out, using `reset` can be dangerous and it is not suitable if you need to be
+more surgical in what you want to change, affecting just what was done on a commit a
+while ago, potentially already shared with collaborators. To address that, we have
+`revert`.
+
+`git revert` creates a commit that exactly cancels the changes made by a previous one.
+It is a new commit, so it is part of the history, but its purpose is to undo something
+done in the past.
+
+The syntax in this case is:
+~~~
+$ git revert --no-edit COMMIT_HASH
```

Comment:
I think I can mention that they can omit the `--no-edit` and add a message, but I rather leave it there as default so there is, at least, a message saying that certain other commit was reverted.

Reply:
If you omit `--no-edit` your text editor will open up with the default message about reverting commit X. It just gives the user an opportunity to say a bit more about why the original commit was problematic.

Reply:
The problem is that the default text editor is often `vim`, and people does not know how to get out of it 😆

Reply:
So the editor's being tricky reasoning was definitely a part of why the original decision was to use --no-edit everywhere. It's particularly difficult because different people will get different editors then you have to deal with every case in the class in order to proceed and if people are a bit behind it gets worse etc. etc. Even more tricky to deal with online because a helper can't just wander over and see what's going on.

I'm not against re-examining this decision if we want to dedicate the time to it. We could probably do the config to set nano as the default editor across platforms which makes things a bit easier and more consistent.

Reply:
How about I open an issue for it? I think we might want to revisit this.

My concerns with avoiding talking about editors are that:

1. Students will complete the intermediate-level course without knowing how to write longer commit messages, which is quite an important thing to know how to do
2. If a student forgets to add the `--no-edit` flag they're going to end up trapped in vim anyway

Reply:
Issue open https://github.com/ImperialCollegeLondon/intermediate_grad_school_git_course/issues/22